File: CodingToUnitTest/StockCalculator.cs
using NUnit.Framework;
using NUnitLite;
using System;
using System.Reflection;


namespace CodepadTestSample
{
    class Program
    {
        static int Main(string[] args)
        {
            return new AutoRun(Assembly.GetCallingAssembly()).Execute(new String[] { "--labels=All" });
        }

        public interface IInventoryRepo
        {
            int GetStock(int productId);

        }

        public class InventoryRepo
        {
            public int GetStock(int productId)
            {
                //This would actually call something like a DB to get the actual stock level
                return new Random().Next(0, 120);
            }
        }

        public class FakeInventoryRepo : IInventoryRepo
        {
            private readonly int _stock;
            public FakeInventoryRepo(int stock)
            {
                _stock = stock;
            }
            public int GetStock(int productId)
            {
                return _stock;
            }
        }

        public class Calculator
        {
            private readonly IInventoryRepo _inventoryRepo;
            public Calculator(IInventoryRepo inventoryRepo)
            {
                this._inventoryRepo = inventoryRepo;
            }

            public double GrossTotal(double price, int quantity) => (price * quantity);

            private const double vatRate = 1.2;
            public double NetTotal(double price, int quantity) => this.GrossTotal(price, quantity) * vatRate;

            public double BulkBuyDiscount(int quantity)
            {

                if (quantity < 100)
                    //No discount
                    return 1;

                if (quantity < 1000)
                    //10 percent
                    return 0.90;

                //a generous 20 percent 
                return 0.8;

            }


            public bool IsStockRunningLow(int productId)
            {
                var currentStock = _inventoryRepo.GetStock(productId);
                return (currentStock < 10);
            }

            public double StockRunningLowMultipler(int productId)
            {
                if (IsStockRunningLow(productId))
                {
                    //add five percent if stock is running low
                    return 1.05;
                }
                return 1;
            }

            public double FinalTotal(int productId, double price, int quantity, bool calculateWithVat)
            {
                var intialTotal = (calculateWithVat)
                                    ? NetTotal(price, quantity)
                                    : GrossTotal(price, quantity);

                return intialTotal * (StockRunningLowMultipler(productId)) * BulkBuyDiscount(quantity);
            }

            public bool IsStockAvailable(int productId, int quantity)
            {
                var currentStock = _inventoryRepo.GetStock(productId);
                return (currentStock >= quantity);
            }

        }


        [TestFixture]
        public class StockCalculatorTests
        {
            private Calculator _cal;
            private IInventoryRepo _fakeInventoryRepo;


            /*
                Tests for GrossTotal, NetTotal and or BulkBuyDiscount
            */

            [Test]
            public void TestGrossTotal()
            {
                _fakeInventoryRepo = new FakeInventoryRepo(1);
                _cal = new Calculator(_fakeInventoryRepo);
                var actual = _cal.GrossTotal(100, 10);
                Assert.AreEqual(1000, actual);
            }

            [Test]
            public void TestNetTotal()
            {
                _fakeInventoryRepo = new FakeInventoryRepo(1);
                _cal = new Calculator(_fakeInventoryRepo);
                var actual = _cal.NetTotal(100, 10);
                Assert.AreEqual(1200, actual);
            }

            [Test]
            public void TestBulkBuyDiscount()
            {
                _fakeInventoryRepo = new FakeInventoryRepo(1);
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(1, _cal.BulkBuyDiscount(50));
                Assert.AreEqual(0.8, _cal.BulkBuyDiscount(2000));
                Assert.AreEqual(0.90, _cal.BulkBuyDiscount(150));
            }


            /*
                Tests for IsStockRunningLow
            */
            [Test]
            public void IsStockRunningLow()
            {
                _fakeInventoryRepo = new FakeInventoryRepo(50);//stock
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(false, _cal.IsStockRunningLow(5));//any productId

                _fakeInventoryRepo = new FakeInventoryRepo(9);
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(true, _cal.IsStockRunningLow(4));
            }


            [Test]
            public void TestStockRunningLowMultipler()
            {
                _fakeInventoryRepo = new FakeInventoryRepo(9);//stock
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(1.05, _cal.StockRunningLowMultipler(4)); //any productId

                _fakeInventoryRepo = new FakeInventoryRepo(50);
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(1, _cal.StockRunningLowMultipler(5));
            }

            /*
                 test for FinalTotal
            */
            [Test]
            public void TestFinalTotal()
            {
                //stock is available,vat is true
                _fakeInventoryRepo = new FakeInventoryRepo(10);//stock
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(1200, _cal.FinalTotal(5, 100, 10, true)); //any productId

                //stock is available, vat is false
                _fakeInventoryRepo = new FakeInventoryRepo(10);
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(1000, _cal.FinalTotal(5, 100, 10, false)); //any productId

                //stock is less, vat is true
                _fakeInventoryRepo = new FakeInventoryRepo(9);
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(1260, _cal.FinalTotal(5, 100, 10, true)); //any productId 1200x1.05x

                //stock is less, vat is false
                _fakeInventoryRepo = new FakeInventoryRepo(9);
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(1050, _cal.FinalTotal(5, 100, 10, false)); //any productId 1000x1.05x
            }

            /*
                Write implemenation and test IsStockAvailable
            */
            [Test]
            public void IsStockAvailable()
            {
                _fakeInventoryRepo = new FakeInventoryRepo(50);//stock
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(true, _cal.IsStockAvailable(5, 20));//any productId,quantity

                _fakeInventoryRepo = new FakeInventoryRepo(50);
                _cal = new Calculator(_fakeInventoryRepo);
                Assert.AreEqual(false, _cal.IsStockAvailable(5, 100));
            }
        }
    }
}

File: CodingToUnitTest/StockCalculator.js
const _ = require('lodash');

class InventoryRepo  {
  GetStock(productId){
    //This would actually call something like a DB to get the actual stock level
    return Math.random() * 120;
  }
}

class Calculator {
  _inventoryRepo;
  constructor(inventoryRepo) {
    this._inventoryRepo = inventoryRepo;
  }

  GrossTotal(price, quanity) {
    return price * quanity;
  }

  vatRate = 1.2;

  NetTotal(price, quanity) {
    return this.GrossTotal(price, quanity) * this.vatRate;
  }

  BulkBuyDiscount(quanity) {
    if (quanity < 100)
      //No discount
      return 1;

    if (quanity < 1000)
      //10 percent
      return 0.9;

    //a generous 20 percent
    return 0.8;
  }

  IsStockRunningLow(productId) {
    var currentStock = this._inventoryRepo.GetStock(productId);
    return currentStock < 10;
  }

  StockRunningLowMultipler(productId) {
    if (this.IsStockRunningLow(productId)) {
      //add five percent if stock is running low
      return 1.05;
    }
    return 1;
  }

  FinalTotal(
    productId,
    price,
    quanity,
    calculateWithVat
  ) {
    const intialTotal = calculateWithVat
      ? this.NetTotal(price, quanity)
      : this.GrossTotal(price, quanity);

    return (
      intialTotal *
      this.StockRunningLowMultipler(quanity) *
      this.BulkBuyDiscount(quanity)
    );
  }

  IsStockAvailable(productId, quanity) {
    throw new Error("not Implemented");
  }
}



const Chia = require("chai");
const sinon = require("sinon");
const sinonChai = require("sinon-chai");

const Mocha = require("mocha");
const assert = require("assert");
const mocha = new Mocha();

// Bit of a hack, but thats how to make it work in code pad
mocha.suite.emit("pre-require", this, "solution", mocha);

Mocha.describe('Calculator', () => {
  let cal;
  let inventoryRepoMock;

  beforeEach(() => {
    inventoryRepoMock = sinon.createStubInstance(InventoryRepo);
    cal = new Calculator(inventoryRepoMock);
  });

  Mocha.it('GrossTotal', () => {
    const price = 50;
    const quantity = 5;
    const result = cal.GrossTotal(price, quantity);
    Chia.expect(result).to.equal(250);
  });

  Mocha.it('NetTotal', () => {
    const price = 50;
    const quantity = 5;
    const expectedGrossTotal = 250;
    const vatRate = 1.2;
    sinon.stub(cal, 'GrossTotal').returns(expectedGrossTotal);

    const result = cal.NetTotal(price, quantity);
    Chia.expect(result).to.equal(expectedGrossTotal * vatRate);
  });

  Mocha.it('BulkBuyDiscount correctly for quantity < 100', () => {
    const quantity = 50;
    const result = cal.BulkBuyDiscount(quantity);
    Chia.expect(result).to.equal(1);
  });

  Mocha.it('BulkBuyDiscount correctly for quantity < 1000', () => {
    const quantity = 500;
    const result = cal.BulkBuyDiscount(quantity);
    Chia.expect(result).to.equal(0.9);
  });

  Mocha.it('BulkBuyDiscount correctly for quantity >= 1000', () => {
    const quantity = 1000;
    const result = cal.BulkBuyDiscount(quantity);
    Chia.expect(result).to.equal(0.8);
  });

  Mocha.it('if stock is running low', () => {
    inventoryRepoMock.GetStock.returns(5);
    const result = cal.IsStockRunningLow('product-id');
    Chia.expect(result).to.be.true;
  });

  Mocha.it('if stock is not running low', () => {
    inventoryRepoMock.GetStock.returns(15);
    const result = cal.IsStockRunningLow('product-id');
    Chia.expect(result).to.be.false;
  });

  Mocha.it('StockRunningLowMultipler correctly when stock is running low', () => {
    sinon.stub(cal, 'IsStockRunningLow').returns(true);
    const result = cal.StockRunningLowMultipler('product-id');
    Chia.expect(result).to.equal(1.05);
  });

  Mocha.it('StockRunningLowMultipler correctly when stock is not running low', () => {
    sinon.stub(cal, 'IsStockRunningLow').returns(false);
    const result = cal.StockRunningLowMultipler('product-id');
    Chia.expect(result).to.equal(1);
  });

  Mocha.it('FinalTotal correctly', () => {
    const price = 100;
    const quantity = 10;
    const productId = 'product-id';
    sinon.stub(cal, 'NetTotal').returns(1200);
    sinon.stub(cal, 'GrossTotal').returns(1000);
    sinon.stub(cal, 'StockRunningLowMultipler').returns(1.05);
    sinon.stub(cal, 'BulkBuyDiscount').returns(0.9);

    const result = cal.FinalTotal(productId, price, quantity, true);
    Chia.expect(result).to.equal(1200 * 1.05 * 0.9);
  });

});

mocha.run();